File: 3264-final-array-state-after-k-multiplication-operations-i/3264-final-array-state-after-k-multiplication-operations-i.cpp
class Solution {
public:
    vector<int> getFinalState(vector<int>& nums, int k, int multiplier) {
        priority_queue<pair<int,int>, vector<pair<int,int>>, greater<pair<int,int>>> pq; 
        for(int i = 0; i < nums.size(); i++){
            pq.push({nums[i],i}); 
        }

        for(int i = 0; i < k; i++){
            pair<int,int> curr = pq.top(); 
            pq.pop(); 

            int num = curr.first, index = curr.second;  
            nums[index] *= multiplier; 
            pq.push({num * multiplier, index}); 
        }

        return nums; 
    }
};

File: 3264-final-array-state-after-k-multiplication-operations-i/3264-final-array-state-after-k-multiplication-operations-i.java
class Solution {
    public int[] getFinalState(int[] nums, int k, int multiplier) {
        PriorityQueue<int[]> heap = new PriorityQueue<>((a,b) -> {
            if(a[0] == b[0]){
                return Integer.compare(a[1],b[1]); 
            }
            return Integer.compare(a[0],b[0]); 
        });

        for(int i = 0; i < nums.length; i++){
            heap.offer(new int[]{nums[i],i}); 
        }

        while(k-- > 0){
            int[] smallest = heap.poll(); 
            nums[smallest[1]] *= multiplier;  
            heap.offer(new int[] {nums[smallest[1]], smallest[1]}); 
        }
        
        return nums; 
    }
}